contrib/babelfishpg_tsql/src/dbcmds.c

```diff
@@ -1225,3 +1258,125 @@ create_guest_schema_for_all_dbs(PG_FUNCTION_ARGS)
 
 	PG_RETURN_INT32(0);
 }
+
+
+/*
+* This function is only being used during upgrade to v4.4.0
+* to create database roles db_accessadmin for each database
+*/
+PG_FUNCTION_INFO_V1(create_database_roles_for_all_dbs);
+Datum
+create_database_roles_for_all_dbs(PG_FUNCTION_ARGS)
+{
+	Relation	sysdatabase_rel;
+	TableScanDesc scan;
+	HeapTuple	tuple;
+	Form_sysdatabases bbf_db;
+	char        *dbname;
+	int         saved_nest_level = 0;
+
+	if (!creating_extension)
+		ereport(ERROR,
+				(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+				 errmsg("%s can only be called from an SQL script executed by CREATE/ALTER EXTENSION",
+						"create_database_roles_for_all_dbs()")));
+
+	saved_nest_level = pltsql_new_guc_nest_level();
+	set_config_option("babelfishpg_tsql.sql_dialect", "tsql",
+						GUC_CONTEXT_CONFIG,
+						PGC_S_SESSION, GUC_ACTION_SAVE, true, 0, false);
+	set_config_option("babelfishpg_tsql.migration_mode", physical_schema_name_exists("dbo") ? "single-db" : "multi-db",
+						GUC_CONTEXT_CONFIG,
+						PGC_S_SESSION, GUC_ACTION_SAVE, true, 0, false);
+
+	sysdatabase_rel = table_open(sysdatabases_oid, RowExclusiveLock);
+	scan = table_beginscan_catalog(sysdatabase_rel, 0, NULL);
+
+	while ((tuple = heap_getnext(scan, ForwardScanDirection)) != NULL)
+	{
+		StringInfoData query;
+		List           *parsetree_list;
+		Node           *stmt;
+		const char     *db_owner;
+		const char     *db_accessadmin;
+		const char     *old_createrole_self_grant = GetConfigOption("createrole_self_grant", false, true);
+		int            i = 0;
+		int save_sec_context = 0;
+		Oid save_userid = InvalidOid;
+
+		bbf_db = (Form_sysdatabases) GETSTRUCT(tuple);
+		dbname = text_to_cstring(&(bbf_db->name));
+		db_owner = get_db_owner_name(dbname);
+		db_accessadmin = get_db_accessadmin_role_name(dbname);
+
+		initStringInfo(&query);
+
+		appendStringInfo(&query, "CREATE ROLE dummy ROLE dummy; ");
+		appendStringInfo(&query, "GRANT CREATE ON DATABASE dummy TO dummy; ");
+
+		parsetree_list = raw_parser(query.data, RAW_PARSE_DEFAULT);
```

Comment:
This can be done only once outside of the while loop. Here we just need to update the rolename/dbname.

Reply:
Done